src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -388,15 +465,37 @@ private bool ShouldInvalidateServer(Exception exception)
                         code = (ServerErrorCode)writeConcernError.GetValue("code", -1).ToInt32();
                         message = writeConcernError.GetValue("errmsg", null)?.AsString;
 
-                        if (IsNotMasterOrRecovering(code, message))
+                        if (IsStateChangeError(code, message))
                         {
-                            return true;
+                            return !IsStaleStateChangeError(commandException.Result, out invalidatingResponseTopologyVersion);
                         }
                     }
                 }
             }
 
+            invalidatingResponseTopologyVersion = null;
             return false;
+
+            bool IsStaleStateChangeError(BsonDocument response, out TopologyVersion nonStaleResponseTopologyVersion)
+            {
+                if (_connectionPool.Generation > connection.Generation)
+                {
+                    // stale generation number
+                    nonStaleResponseTopologyVersion = null;
+                    return true;
+                }
+
+                var responseTopologyVersion = TopologyVersion.FromMongoCommandResponse(response);
+                // We use FresherThanOrEqualTo instead of FresherThan because a state change should come with a new
+                // topology version.
+                // We cannot use StalerThan(responseTopologyVersion, description.TopologyVersion) because due to how
+                // TopologyVersions comparisons are defined, FresherThanOrEqualTo(x, y) does not imply StalerThan(y, x)
+                bool isStale = TopologyVersion.IsFresherThanOrEqualTo(description.TopologyVersion, responseTopologyVersion);
```

Comment:
I find it very confusing that a variable called `isStale` is initialized by calling a function with a name containing "Fresher".

Can we not use:

    var isStale = TopologyVersion.IsStalerThanOrEqualTo(responseTopologyVersion, description.TopologyVersion);